### DISCUSSION TITLE
getSession() call doesn't seem to update useSession()

### DISCUSSION THREAD

**Discussion by atawfique21** [nextauthjs/next-auth#2380] [Help]
I'm trying to update the session when a user changes username with a button click. 

From research done, I've learned that: I **must** use Database with JWT turned off as you mentioned: 
> When using JWT, the data is stored in a cookie, and you have to trigger the sign-in flow again to get the new user data.

What I am trying to achieve now is mentioned in #2269: 
> If you use a database, you can update the data directly, and if you configured it correctly, I believe the getSession call should update it already. 

Currently, I have an API call. When the button is clicked, I update a custom column in the users table generated by next-auth. I have updated the model so all of this works. 

```javascript
import { useSession, getSession } from 'next-auth/client'
import axios from "axios"


function newUser() {
  let [session, loading] = useSession()

  const handleUpdate = async () => {
    console.log(session) // this returns the session where session.user.custom_username is old value
    axios.get('http://localhost:3000/api/update-username').then(async () => {
      console.log(await getSession()) // this returns the session where session.user.custom_username is new value
    })
  }

  return (
    <div>
      <p>Hi there, you are new around here</p>
      <button onClick={() => handleUpdate()}>change username</button>
      <h1>{session && session.user.custom_username}</h1>
      {/* No matter what, session.user.custom_username always remains the same until I refresh the page */}
    </div>
  )
}

export default newUser;
```

I understand that I can force the page to refresh with Router but I'm ideally looking for a way to update Session in this component when this Axios call is completed. 

So far: 
- [x] On page load, session.user.custom_username shows current DB value
- [x] On button click, axios updates the custom username in the backend
- [x] In the .then() part of the Axios call, getSession() returns the new custom_username value as part of the session
- [ ] Session is updated for newUser() component and the `<h1>` changes it's value appropriately. 

Please let me know what I'm missing or doing wrong.
Thanks for your work and time.

**Comment by atawfique21** [comment]
I believe I'm come to a bit of a workaround. 
This workaround consists of ditching `useSession()` as I can't find a way to update the session value inside of it and instead using `getSession()` to set a state value using `useEffect()`

Can anyone let me know if this is the best way to approach this problem?

```JavaScript
import { getSession } from 'next-auth/client'
import { useEffect, useState } from 'react'
import axios from "axios"


function newUser() {
  let [realSession, setRealSession] = useState()

  useEffect(() => {
    async function fetch() {
      let session = await getSession()
      setRealSession(session)
    }

    fetch()
  }, [])

  const handleUpdate = async () => {
    axios.get('http://localhost:3000/api/update-username').then(async () => {
      let newSession = await getSession()
      setRealSession(newSession)
    })
  }

  return (
    <div>
      <p>Hi there, you are new around here</p>
      <button onClick={() => handleUpdate()}>change username</button>
      <h1>{realSession && realSession.user. custom_username}</h1>
    </div>
  )
}

export default newUser;
```

**Comment by balazsorban44** [comment]
Thank you for opening this discussion! :pray:

I actually don't understand why people are tripping on this. :confused: 

When `getSession` is called client-side, it is supposed to send a message that should re-trigger the session fetching in `useSession`:

https://github.com/nextauthjs/next-auth/blob/2155c93a3ceeaa38a9d4caf1e109dea3ff0b6219/src/client/index.js#L161-L167

https://github.com/nextauthjs/next-auth/blob/2155c93a3ceeaa38a9d4caf1e109dea3ff0b6219/src/client/index.js#L82-L159

Granted, the current solution is not really elegant and ignores some basic React rules, I am not surprised if we have a weird bug that makes this not work... Hopefully, the new version of the client code that is more React oriented will be more stable in this manner.

Have a glimpse at it here: https://github.com/nextauthjs/next-auth/blob/next/src/client/react.js

I still hope an `update` method or similar exposed on `useSession` will make the most sense here, but I am open to suggestions for an alternative API in #2269

**Reply by atawfique21** [reply]
To be honest, I think this is a shortcoming of next-auth for the time being. We really do need a robust method to deal with users changing their attributes like usernames, profile pictures, etc. And upon those attributes changing, sending the session token down the prop line to be re-rendered by React. In my 12+ hour hunt, I've seen countless discussions/issue tickets + stackOverflow questions and all of the solutions (including my solutions!), in my opinion, are very hacky and defeat optimization benefits next-auth otherwise provides. 

I certainly think your solution of an `update` method on `useSession` is the way to go. This allows for fine-tuning and users like me would be able to ditch the aforementioned broadcast not working. Of course, when using `update`, I would expect all components that are using `useSession` to get the new session object for no-refresh rendering. 

For the time being, I'm thinking the best approach is to use React Router and refresh the page upon submit and forgo my 'workaround' where I ditch `useSession()` for `getSession()`. I come to this because my `<Header>` component contains a reference to the user's custom username and that component would also need to forgo `useSession()` for `getSession()`. And on top of that, I'm not even sure how to force `<Header>` to re-render when `<newUser>` is the one that submits these changes. It seems like when updating user attributes client-side, for the time being, a page refresh is just the way to go. Do you agree with this sentiment or perhaps you have a better way of approaching this while we wait for #2269 to be coded & merged? 

On the bright side: At least a page refresh would only need to occur only when user something about their profile, which isn't very often in my application.
